roscala/src/main/scala/coop/rchain/roscala/prim/fixnum.scala

```diff
@@ -13,4 +21,369 @@ object fixnum {
+    override val maxArgs: Int = 1
+
+    @checkTypeMismatch[Fixnum]
+    override def fn(ctxt: Ctxt, globalEnv: GlobalEnv): Ob = {
+      val m = ctxt.argvec.value.head.asInstanceOf[Fixnum]
+
+      Fixnum(~m.value)
+    }
+  }
+
+  object fxMdiv extends Prim {
+    override val name: String = "fx-mdiv"
+    override val minArgs: Int = 0
+    override val maxArgs: Int = MaxArgs
+
+    @checkTypeMismatch[Fixnum]
+    override def fn(ctxt: Ctxt, globalEnv: GlobalEnv): Ob = {
+      val n = ctxt.nargs
+
+      val commonBits = ctxt.argvec.value.take(n).foldLeft(Fixnum(~0)) {
+        case (accum, fixnum: Fixnum) =>
+          accum & fixnum
+      }
+
+      if (commonBits.value != 0) {
+        Fixnum(1)
+      } else {
+        Fixnum(0)
+      }
+    }
+  }
+
+  object fxCdiv extends Prim {
+    override val name: String = "fx-cdiv"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[Fixnum]
+    override def fn(ctxt: Ctxt, globalEnv: GlobalEnv): Ob = {
+      val m = ctxt.argvec.value.head.asInstanceOf[Fixnum]
+      val n = ctxt.argvec.value(1).asInstanceOf[Fixnum]
+
+      try {
+        Fixnum(Math.ceil(m.value.toDouble / n.value.toDouble).toInt)
+      } catch {
+        case _: ArithmeticException =>
+          Prim.runtimeError(ctxt, "arithmetic exception")
+      }
+    }
+  }
+
+  object fxAsl extends Prim {
+    override val name: String = "fx-asl"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[Fixnum]
+    override def fn(ctxt: Ctxt, globalEnv: GlobalEnv): Ob = {
+      val m = ctxt.argvec.value.head.asInstanceOf[Fixnum]
+      val n = ctxt.argvec.value(1).asInstanceOf[Fixnum]
+
+      m << n
+    }
+  }
+
+  object fxAsr extends Prim {
+    override val name: String = "fx-asr"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[Fixnum]
+    override def fn(ctxt: Ctxt, globalEnv: GlobalEnv): Ob = {
+      val m = ctxt.argvec.value.head.asInstanceOf[Fixnum]
+      val n = ctxt.argvec.value(1).asInstanceOf[Fixnum]
+
+      m >> n
+    }
+  }
+
+  object fxLsl extends Prim {
+    override val name: String = "fx-lsl"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[Fixnum]
+    override def fn(ctxt: Ctxt, globalEnv: GlobalEnv): Ob = {
+      val m = ctxt.argvec.value.head.asInstanceOf[Fixnum]
+      val n = ctxt.argvec.value(1).asInstanceOf[Fixnum]
+
+      m << n
+    }
+  }
+
+  object fxLsr extends Prim {
+    override val name: String = "fx-lsr"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkTypeMismatch[Fixnum]
+    override def fn(ctxt: Ctxt, globalEnv: GlobalEnv): Ob = {
+      val m = ctxt.argvec.value.head.asInstanceOf[Fixnum]
+      val n = ctxt.argvec.value(1).asInstanceOf[Fixnum]
+
+      m >>> n
+    }
+  }
+
+  private def checkFixnum(ctxt: Ctxt, n: Int, elem: Array[Ob]): Ob =
```

Comment:
Can you give me some advise of how to implement the checkFixnum?

Reply:
It looks correct to me.
If `elem(n)` is no `Fixnum` it does some error handling stuff and then returns `DEADTHREAD`.
If not, it returns `elem(n)` as a `Fixnum`.